### PR TITLE
Get neighbors

### PR DESCRIPTION
**Pull request by ikarabulut**
I added a few new pieces of functionality to get this Story completed.

In my Board Class
- setAliveAt() : allows you to set a cell to alive
     - Within this method we call the private method `getCell()` which finds the row and then cell within that row.
- isCellAliveAt() : allows you to see if a certain cell is alive, it returns a boolean
     - This method is used for testing but will be needed once we start adding the rules
-getAliveNeighborsCountOf() : returns an int of how many alive neighbors there are. This implements a few private methods

Private methods
- getNeighborsOf() : This returns a List of the 8 neighbors around the specified cell and then passes that to getAliveNeighborsCountOf()
-**I separated out each neighbor into a different private method, this breaks apart the complex implementation methods to do wrap-around functionality, especially for corner cells**

Tests were added to ensure the count of alive neighbors is correct and the wrap around works correctly

### CODE REVIEW

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 1]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
         assertFalse(isDeadBoard);
     }
 
-}
\ No newline at end of file
+    @Test
+    @DisplayName("A cell that is dead can be set to alive")
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:73]:
```suggestion
    @DisplayName("The board can bring a cell to life at a given location")
```

Just trying to differentiate between the board being able to tell a cell to come alive and the feature that will be in the Cell class, where it can bring itself to life.

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 2]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+
+        board.setAliveAt(1, 1);
+        cellStatus = board.isCellAliveAt(1, 1);
+
+        assertTrue(cellStatus);
+
+    }
+
+    @Nested
+    @DisplayName("Tests for getAliveNeighborsCountOf")
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:88]:
```suggestion
    @DisplayName("getAliveNeighborsCountOf()")
```

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 3]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+    class GetAliveNeighborsCountOf {
+        public Board board;
+        @BeforeEach
+        void createBoard() {
+            board = new Board(3,3);
+        }
+        @Test
+        @DisplayName("A cell with 0 alive neighbor should return 0 alive neighbor")
+        void getAliveNeighborsCountOf() {
+            int numOfAliveNeighbors = board.getAliveNeighborsCountOf(1,1);
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:98]: Not exactly sure what to do (if anything) at the moment, but just wanted to point out that this spec relies on the fact that creating a new instance of a `Board` creates a board with all "dead" cells. If we were to change that, this spec would fail.

One solution would be to have a helper method in the specs where you set every cell to be "dead" and return a new board.

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 4]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+
+            int numOfAliveNeighbors = board.getAliveNeighborsCountOf(1,1);
+
+            assertEquals(numOfAliveNeighbors, 2);
+        }
+
+        @Test
+        @DisplayName("A cell with 8 alive neighbors should return 8 alive neighbors")
+        void getAliveNeighborsCountOfTest3() {
+            board.setAliveAt(0,0);
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:126]: Similar to above, if you find yourself doing this more than twice in the specs, I'd extract this to a helper method - `setEntireBoardAlive()` that iterates over any size board and sets every cell to be "alive"

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 5]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+            board.setAliveAt(2,1);
+            board.setAliveAt(2,2);
+
+            int numOfAliveNeighbors = board.getAliveNeighborsCountOf(1, 1);
+
+            assertEquals(numOfAliveNeighbors, 8);
+        }
+
+        @Test
+        @DisplayName("An alive cell on the left edge should register as an alive cell for a right edge cell")
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:141]:
```suggestion
        @DisplayName("should count neighbors that wrap from the right side of a row to the left side")
```

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 6]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+            Board largerBoard = new Board(5,5);
+            largerBoard.setAliveAt(2, 0);
+
+            int numOfAliveNeighbors = largerBoard.getAliveNeighborsCountOf(2, 4);
+
+            assertEquals(numOfAliveNeighbors, 1);
+        }
+
+        @Test
+        @DisplayName("An alive cell on the top left edge should register as an alive cell for a bottom right edge")
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:152]:
```suggestion
        @DisplayName("should count neighbors that wrap diagonally from the top left to the bottom right of the board")
```

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 7]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+            Board largerBoard = new Board(5, 5);
+            largerBoard.setAliveAt(0,0);
+
+            int numOfAliveNeighbors = largerBoard.getAliveNeighborsCountOf(4, 4);
+
+            assertEquals(numOfAliveNeighbors, 1);
+        }
+
+        @Test
+        @DisplayName("An alive cell on the bottom edge should register as an alive cell for a top edge cell")
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:163]:
```suggestion
        @DisplayName("should count neighbors that wrap from the top of a column to the bottom")
```

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 8]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+        @Test
+        @DisplayName("A cell with 0 alive neighbor should return 0 alive neighbor")
+        void getAliveNeighborsCountOf() {
+            int numOfAliveNeighbors = board.getAliveNeighborsCountOf(1,1);
+            assertEquals(numOfAliveNeighbors, 0);
+        }
+
+        @Test
+        @DisplayName("A cell with 1 alive neighbor should return 1 alive neighbor")
+        void getAliveNeighborsCountOfTest1() {
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:104]: We should probably use a more descriptive name here. I'm not sure what the standard practice is for this format. This applies to all the test method names.

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 9]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+
+            int numOfAliveNeighbors = board.getAliveNeighborsCountOf(1, 1);
+
+            assertEquals(numOfAliveNeighbors, 8);
+        }
+
+        @Test
+        @DisplayName("An alive cell on the left edge should register as an alive cell for a right edge cell")
+        void getAliveNeighborsCountOfTest4() {
+            Board largerBoard = new Board(5,5);
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:143]: Why the switch to using a bigger board here?

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 10]
```diff
@@ -69,4 +69,121 @@ void setRandomGrid() {
...
+            Board largerBoard = new Board(5, 5);
+            largerBoard.setAliveAt(4, 2);
+
+            int numOfAliveNeighbors = largerBoard.getAliveNeighborsCountOf(0, 2);
+
+            assertEquals(numOfAliveNeighbors, 1);
+        }
+
+        @Test
+        @DisplayName("An alive cell on the bottom left edge should register as an alive cell for a top right cell")
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/BoardTest.java:174]:
```suggestion
        @DisplayName("should count neighbors that wrap diagonally from the bottom left to the top right of the board")
```

File: app/src/main/java/com/ikarabulut/Board.java [hunk 11]
```diff
@@ -22,6 +23,7 @@ public Board(int numOfRows, int numOfColumns) {
         grid = setGrid();
     }
 
+    // PUBLIC METHODS
     public List<ArrayList> setGrid() {
```
Comment (riccjohn) [inline app/src/main/java/com/ikarabulut/Board.java:27]: I still think we're generating a board here, not setting a grid, but let's see what Jeff & Claudia think too.
```suggestion
    public List<ArrayList> generate() {
```

Similarly, I'd rename `setRandomGrid()` to `generateRandomBoard()` and `getGrid() to `getBoard()`

File: app/src/main/java/com/ikarabulut/Board.java [hunk 12]
```diff
@@ -49,4 +51,135 @@ public void setRandomGrid() {
...
+        Cell selectedCell = getCell(rowNum, colNum);
+        selectedCell.reviveMe();
+    }
+
+    public boolean isCellAliveAt(int rowNum, int colNum) {
+        Cell selectedCell = getCell(rowNum, colNum);
+        return selectedCell.checkIfAlive();
+    }
+
+    public int getAliveNeighborsCountOf(int rowNum,int colNum) {
```
Comment (riccjohn) [inline app/src/main/java/com/ikarabulut/Board.java:65]:
```suggestion
    public int getNumberOfAliveNeighbors(int rowNum,int colNum) {
```

**Review by riccjohn — CHANGES_REQUESTED** [review]
This is looking good!! You're definitely taking a lot of the learnings from the Ruby version and applying them here, which is great.

**BONUS COMMENT!**: I couldn't comment on this bc it's not part of the PR, but I'd also rename `Cell.reviveMe` to just `Cell.revive`

**Comment by riccjohn** [comment]
another Cell class comment:
- Is it possible to have the `checkIfAlive` method be named `isAlive` or `isAlive?` (not sure if the `?` is allowed in Java method names). I know the instance variable is named `isAlive`, but hopefully it's ok to have a public getter method for that.

File: app/src/main/java/com/ikarabulut/Board.java [hunk 13]
```diff
@@ -49,4 +51,135 @@ public void setRandomGrid() {
...
+        Cell topLeftNeighbor;
+        if (rowNum == 0 && colNum == 0) {
+            topLeftNeighbor = getCell(rows - 1, columns - 1);
+        } else if (rowNum == 0) {
+            topLeftNeighbor = getCell(rows - 1,colNum - 1);
+        } else {
+            topLeftNeighbor = getCell(rowNum - 1, colNum - 1);
+        }
+        return topLeftNeighbor;
+    }
```
[the remark this answers is not in the document]
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/Board.java:103]: Thanks for pointing this out. i was confused at first with IntelliJ cause they sorta add a 'fake' line between the methods to show usage but its not a real space.

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 14]
```diff
@@ -3,8 +3,8 @@
 import java.util.ArrayList;
 import java.util.List;
 
-import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.*;
```
[the remark this answers is not in the document]
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:6]: I think that was added initially but I will remove it so I can track which ones I am using in general for my learning. My next merge will show these changes

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 1 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:73]: I like this naming better. Will include this change in my next merge

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 2 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:88]: I kinda like leaving 'Tests for' cause it makes it more easily readable in the test output 
Screen Shot 2022-05-05 at 4 45 51 PM

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 3 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:98]: Yeah this is a good call. in my next merge to this branch you'll see the setDeadBoard() helper method that is run before every test in this nested test suite.

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 4 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:126]: Added this helper as well

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 8 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:104]: Id have to look into this more cause with the display names it looks like you usually just put the name of the method but we cannot use the same method multiple times. Im sure there's some nested testing feature that I have not discovered yet that does this. Ill add a card for this

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 9 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:143]: Cause it will read the alive cell in general if it has one. If I make it a larger board, the only alive cell it will have is the one from the wrap around.

File: app/src/main/java/com/ikarabulut/Board.java (hunk 12 above)
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/Board.java:65]: Better name. I updated it

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 8 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java]: It looks like I should still follow naming conventions. There are a few but I went with [MethodName_StateUnderTest_ExpectedBehavior] which is shown in my most recent push to this branch

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 2 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:91]: Actually its obvious that these are tests so there is no need to call that out. Ill update it.

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 15]
```diff
@@ -69,4 +72,137 @@ void setRandomGrid() {
         assertFalse(isDeadBoard);
     }
 
-}
```
Comment (jramnani) [inline app/src/test/java/com/ikarabulut/BoardTest.java:72]: Note: When you see this symbol in a code diff, it means your text editor did not add a newline character `\n` at the end of the file.  This can make files harder to process by automated analysis tools. So, it is a best practice to configure your text editor to add them for you when you edit a file.  Normally a UNIX editor will do this for you, like [vim does](https://vi.stackexchange.com/questions/3298/how-to-make-vim-automatically-add-a-newline-to-the-end-of-a-file).  But not all editors do this and need to be configured to do so.

Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:72]: Good to know. I guess I should have looked through the file myself to even catch those. I updated my IntelliJ settings to add the newline. On my local file it looks like it added the newlines but git is saying I am still working a clean branch so I can't push those changes. Maybe when I have my next commit it will update

File: app/src/main/java/com/ikarabulut/Board.java (hunk 11 above)
Comment (jramnani) [reply app/src/main/java/com/ikarabulut/Board.java:27]: Prefer to keep domain terms distinct. Using different terms for the same thing leads to code that is confusing to read.  Is it a Board or a Grid?

Also, I think John is right in that generate is a better description for what `setGrid` is doing.

When I read Java code, I expect the following conventions:

* A method name with the prefix of `get` (`getFoo`) returns a field on the object named 'foo'.
* A method name with a prefix of `set` (`setFoo(value)`) sets the field `foo` to the value.

Together these conventions are known as "getters" and "setters".  

Technically the `setGrid` function is setting a value for a field named `grid`.  But it feels like it's doing a lot more than a little validation on an input value.  Also, it's not taking a value to set anything to.  This is why the word `generate` is better for describing what this method is doing.

File: app/src/main/java/com/ikarabulut/Board.java [hunk 16]
```diff
@@ -49,4 +51,142 @@ public void setRandomGrid() {
     public List<ArrayList> getGrid() {
         return grid;
     }
+
+    public void setAliveAt(int rowNum, int colNum) {
```
Comment (jramnani) [inline app/src/main/java/com/ikarabulut/Board.java:55]: One thing that is nice about statically typed languages that is different from dynamic ones is that variables are annotated with their type. In this case `rowNum` and `colNum` are both of type `int`.  That means we can infer that they are, in fact, numbers.  That makes the `Num` suffix for these variable names redundant.

This is a common thing people do at first, because they are used to dynamic languages where they may not know the type of a variable by looking at it.  Adding an object's type to it's name is called [Hungarian Notation](https://en.wikipedia.org/wiki/Hungarian_notation) and is, in general, discouraged because it makes variables harder to read.

File: app/src/main/java/com/ikarabulut/Cell.java [hunk 17]
```diff
@@ -14,7 +14,7 @@ public Cell(){
     public boolean checkIfAlive() {
         return isAlive;
     }
-    public void reviveMe() {
+    public void revive() {
```
Comment (jramnani) [inline app/src/main/java/com/ikarabulut/Cell.java:17]: 👍

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 18]
```diff
@@ -1,18 +1,21 @@
...
-import org.junit.jupiter.api.DisplayName;
 import static org.junit.jupiter.api.Assertions.*;
 
 
 class BoardTest {
 
     @Test
     @DisplayName("Given no arguments, a 10x10 board is created by default")
-    void setGridDefault() {
+    void setGrid_NoArgumentsPassed_10x10BoardIsGenerated() {
```
Comment (jramnani) [inline app/src/test/java/com/ikarabulut/BoardTest.java:18]: Thanks for improving these test names. Also these test DispayNames are cool, and help me understand what condition is being tested for.

Is `@DisplayName` new in JUnit 5?

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 19]
```diff
@@ -69,4 +72,137 @@ void setRandomGrid() {
...
\ No newline at end of file
+    @Test
+    @DisplayName("The board can bring a cell to life at a given location")
+    void setAliveAt_When1_1IsPassed_ReturnsTheCellObjectAt1_1() {
+        Board board = new Board();
+
+        Boolean cellStatus = board.isCellAliveAt(1, 1);
+        assertFalse(cellStatus);
+
+        board.setAliveAt(1, 1);
```
Comment (jramnani) [inline app/src/test/java/com/ikarabulut/BoardTest.java:83]: I like that we send messages to Board to take care of which cells are alive at a given location. Board's job seems to be "Manage a collection of Cells", rather than "I'm a bag of Cells, here you go. You take them."

This prevents Law of Demeter violations later, and makes Board easier to collaborate with for other objects.

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 20]
```diff
@@ -69,4 +72,137 @@ void setRandomGrid() {
...
+        cellStatus = board.isCellAliveAt(1, 1);
+
+        assertTrue(cellStatus);
+
+    }
+
+    @Nested
+    @DisplayName("getNumberOfAliveNeighbors()")
+    class GetNumberOfAliveNeighbors {
+        private void setDeadBoard() {
```
Comment (jramnani) [inline app/src/test/java/com/ikarabulut/BoardTest.java:93]: Nice use of helper methods to make testing easier.

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 21]
```diff
@@ -69,4 +72,137 @@ void setRandomGrid() {
...
+        @DisplayName("A cell with 0 alive neighbor should return 0 alive neighbor")
+        void getNumberOfAliveNeighbors_WhenNoNeighborsAreAlive_Returns0() {
+            int numOfAliveNeighbors = board.getNumberOfAliveNeighbors(1,1);
+            assertEquals(numOfAliveNeighbors, 0);
+        }
+
+        @Test
+        @DisplayName("A cell with 1 alive neighbor should return 1 alive neighbor")
+        void getNumberOfAliveNeighbors_When1NeighborIsAlive_Returns1() {
+            board.setAliveAt(0,1);
```
Comment (jramnani) [inline app/src/test/java/com/ikarabulut/BoardTest.java:125]: It's common practice to use a space in between multiple arguments after the comma. e.g. `board.setAlive(0, 1)`

I notice it's being done at some places, but not consistently everywhere like in this example.

File: app/src/test/java/com/ikarabulut/BoardTest.java [hunk 22]
```diff
@@ -69,4 +72,137 @@ void setRandomGrid() {
...
+            largerBoard.setAliveAt(4, 2);
+
+            int numOfAliveNeighbors = largerBoard.getNumberOfAliveNeighbors(0, 2);
+
+            assertEquals(numOfAliveNeighbors, 1);
+        }
+
+        @Test
+        @DisplayName("Should count neighbors that wrap from the top right of a column to the bottom left")
+        void getNumberOfAliveNeighbors_WhenCellIsTopRight_ReturnsAliveNeighborsFromBottomLeftEdge() {
```
Comment (jramnani) [inline app/src/test/java/com/ikarabulut/BoardTest.java:191]: I notice we're testing that the top edge wraps to the bottom edge. Are we also testing that the bottom edge wraps to the top?  I know world wrapping is the trickiest bit, so it might help to have tighter test coverage around that algorithm.

**Review by jramnani — APPROVED** [review]
Looking good!  I've left some comments and suggestions.

I'd like to see some tighter test coverage around world-wrapping since I know it's the trickiest bit of GoL to get right.  But, I would accept that in a future update.  

There are also some formatting cleanup that would be best to fix before finally merging.

Otherwise, no blockers on my end.  Nice work!

File: app/src/main/java/com/ikarabulut/Board.java (hunk 11 above)
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/Board.java:27]: Good points all around. Updated to board, and generate. Shown in my next push to this branch

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 18 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:18]: Yes! Junit 5 added this feature. I still think my test method names are a bit redundant now but they are still following the naming conventions I am seeing out in the wild

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 21 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:125]: When I add arguments or pass parameters I literally say 'remember Jeff strongly dislikes missing spaces" Im surprised I let a few past. But I went back in and made the proper fixes

File: app/src/test/java/com/ikarabulut/BoardTest.java (hunk 22 above)
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/BoardTest.java:191]: I am testing both corners but there are a few more cases that could be tested.

**Review by riccjohn — APPROVED** [review]
Great work!
